HIVE-26789: Add UserName in CallerContext for queries.

ayushtkn opened the pull request:
### What changes were proposed in this pull request?

Add UserName in CallerContext for HDFS Audit logs to pick.

### Why are the changes needed?

With Impersonation turned off, every call shows from Hive User, So, to allow tracking the actual user.

### Does this PR introduce _any_ user-facing change?

Nopes,

### How was this patch tested?

Tried a Query in Actual Cluster.

cnauroth commented on service/src/java/org/apache/hive/service/cli/operation/Operation.java:
```diff
@@ -237,7 +238,9 @@ protected void createOperationLog() {
    */
   protected void beforeRun() {
-    ShimLoader.getHadoopShims().setHadoopQueryContext(queryState.getQueryId());
+    CallerContext.setCurrent(new CallerContext.Builder("Check").build());
```
Should this line be removed? Unless I'm mistaken, the call to `setHadoopQueryContext` on the next line will overwrite the value set by this line.

ayushtkn replied:
Aah, my bad. Came through my initial dev testing & I forgot to remove it :-(

cnauroth commented on service/src/java/org/apache/hive/service/cli/operation/SQLOperation.java:
```diff
@@ -328,7 +328,8 @@ public Object run() throws HiveSQLException {
           }
-          ShimLoader.getHadoopShims().setHadoopQueryContext(queryState.getQueryId());
+          ShimLoader.getHadoopShims()
+              .setHadoopQueryContext(queryState.getQueryId() + " User:" + parentSessionState.getUserName());
```
For consistency with the other call sites setting query context, please add a space after the colon:

```
... + " User: " + ...`
```

(However, also see my other comment about whether or not we should use spaces. Whatever is decided for the format should be consistent at all call sites.)

ayushtkn replied:
I removed the space itself,

cnauroth commented on cli/src/java/org/apache/hadoop/hive/cli/CliDriver.java:
```diff
@@ -250,7 +250,8 @@ CommandProcessorResponse processLocalCmd(String cmd, CommandProcessor proc, CliS
         // Set HDFS CallerContext to queryId and reset back to sessionId after the query is done
-        ShimLoader.getHadoopShims().setHadoopQueryContext(qp.getQueryState().getQueryId());
+        ShimLoader.getHadoopShims()
+            .setHadoopQueryContext(qp.getQueryState().getQueryId() + " User: " + ss.getUserName());
```
I wonder if we should avoid embedding spaces in the format. Prior usage of caller context that I've seen uses an underscore-delimited format. The Hadoop compatibility guidelines state that the HDFS audit log format should be considered public and stable:

https://hadoop.apache.org/docs/current/hadoop-project-dist/hadoop-common/Compatibility.html#Audit_Log_Output

Embedding spaces could break existing scripts that perform positional parsing using utilities like `cut` and `awk`.

ayushtkn replied:
Thanx for the pointers, no specific reason for using space, if it can break parsers, better to avoid!!

cnauroth reviewed (commented):
@ayushtkn , thanks for the patch. It looks useful! I entered a few questions.

ayushtkn commented:
Thanx @cnauroth for the review, Have addressed the review comments, my initial use case was around the queries only so ignored setting user in the other places, I have now added everywhere

cnauroth reviewed (approved):
+1 (non-binding)

Thank you for incorporating the code review feedback, @ayushtkn .

deniskuzZ commented on cli/src/java/org/apache/hadoop/hive/cli/CliDriver.java:
```diff
@@ -250,12 +250,13 @@ CommandProcessorResponse processLocalCmd(String cmd, CommandProcessor proc, CliS
         // Set HDFS CallerContext to queryId and reset back to sessionId after the query is done
-        ShimLoader.getHadoopShims().setHadoopQueryContext(qp.getQueryState().getQueryId());
+        ShimLoader.getHadoopShims()
+            .setHadoopQueryContext(qp.getQueryState().getQueryId() + "_User:" + ss.getUserName());
```
can we extract "_User:" literal into a constant or have a pattern for String.format() ?

ayushtkn replied:
Changed

deniskuzZ reviewed (commented):
LGTM, minor refactor ask
